### PR TITLE
Fix part #180: Hints and solutions Domain layer

### PR DESCRIPTION
**Pull request by veena14cs**
## Explanation
PR for hints and solution is split into 2 parts.

1) #865 PR aims at introducing hints and solutions parts at domain layer. 
For testing purpose I have added dummy hints in TEST_EXPLORATION_ID_5 . 
2) #868 PR aims at introducing hints and solutions parts at App layer module. 

### Design Doc
https://docs.google.com/document/d/1NrZOTcVXhJxwCG_3nGcp47jBHLufLRrU5Kcv9aVjj0M/edit?usp=sharing

### CODE REVIEW

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt [hunk 1]
```diff
@@ -149,6 +150,37 @@ class ExplorationProgressController @Inject constructor(
...
+          "Cannot submit an answer while the exploration is being loaded."
+        }
+        check(explorationProgress.playStage != PlayStage.SUBMITTING_ANSWER) {
+          "Cannot submit an answer while another answer is pending."
+        }
+        lateinit var hint: Hint
+        try {
+          explorationProgress.stateDeck.submitHintRevealed(state, hintIsRevealed, hintIndex)
+
+          hint = explorationProgress.stateGraph.computeHintForResult(explorationProgress.stateDeck.getCurrentUpdatedEphemeralState().state, hintIsRevealed, hintIndex)
```
Comment (rt4914) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Nit for this line:
```
            hint = explorationProgress.stateGraph.computeHintForResult(
            explorationProgress.stateDeck.getCurrentUpdatedEphemeralState().state,
            hintIsRevealed,
            hintIndex
          )
```

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt [hunk 2]
```diff
@@ -149,6 +150,37 @@ class ExplorationProgressController @Inject constructor(
...
+
+          asyncDataSubscriptionManager.notifyChangeAsync(CURRENT_STATE_DATA_PROVIDER_ID)
+
+        } finally {
+
+        }
+
```
Comment (rt4914) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt:216]: Remove this empty line

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt [hunk 3]
```diff
@@ -149,6 +150,37 @@ class ExplorationProgressController @Inject constructor(
...
+        lateinit var hint: Hint
+        try {
+          explorationProgress.stateDeck.submitHintRevealed(state, hintIsRevealed, hintIndex)
+
+          hint = explorationProgress.stateGraph.computeHintForResult(explorationProgress.stateDeck.getCurrentUpdatedEphemeralState().state, hintIsRevealed, hintIndex)
+
+          asyncDataSubscriptionManager.notifyChangeAsync(CURRENT_STATE_DATA_PROVIDER_ID)
+
+        } finally {
+
```
Comment (rt4914) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt:210]: Remove this clause if not needed

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt [hunk 4]
```diff
@@ -149,6 +150,37 @@ class ExplorationProgressController @Inject constructor(
...
+          "Cannot submit an answer while another answer is pending."
+        }
+        lateinit var hint: Hint
+        try {
+          explorationProgress.stateDeck.submitHintRevealed(state, hintIsRevealed, hintIndex)
+
+          hint = explorationProgress.stateGraph.computeHintForResult(explorationProgress.stateDeck.getCurrentUpdatedEphemeralState().state, hintIsRevealed, hintIndex)
+
+          asyncDataSubscriptionManager.notifyChangeAsync(CURRENT_STATE_DATA_PROVIDER_ID)
+
```
Comment (rt4914) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt:203]: Remove this blank line

File: model/src/main/proto/exploration.proto [hunk 5]
```diff
@@ -108,14 +108,29 @@ message Solution {
   string interaction_id = 1;
   // Flag that is true if correct_answer is the only correct answer of the question.
   bool answer_is_exclusive = 2;
-  string correct_answer = 3;
+
+  CorrectAnswer correct_answer = 3;
```
Comment (rt4914) [inline model/src/main/proto/exploration.proto:112]: Add documentation for each of these fields

File: model/src/main/proto/exploration.proto [hunk 6]
```diff
@@ -108,14 +108,29 @@ message Solution {
...
+  bool solution_is_revealed = 5;
 }
 
+// Structure for a Correct answer in Solution
+// Maps from: data/src/main/java/org/oppia/data/backends/gae/model/GaeSolution.kt
+message CorrectAnswer {
+  int32 denominator = 1;
+  int32 numerator = 2;
+  int32 wholeNumber = 3;
+  bool isNegative = 4;
```
Comment (rt4914) [inline model/src/main/proto/exploration.proto]: is_negative

File: model/src/main/proto/exploration.proto [hunk 7]
```diff
@@ -108,14 +108,29 @@ message Solution {
...
+
+  bool solution_is_revealed = 5;
 }
 
+// Structure for a Correct answer in Solution
+// Maps from: data/src/main/java/org/oppia/data/backends/gae/model/GaeSolution.kt
+message CorrectAnswer {
+  int32 denominator = 1;
+  int32 numerator = 2;
+  int32 wholeNumber = 3;
```
Comment (rt4914) [inline model/src/main/proto/exploration.proto]: whole_number

**Review by rt4914 — APPROVED** [review]
Nit changes suggested otherwise Looks good to me.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt (hunk 2 above)
Comment (nikitamarysolomanpvt) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt:216]: +1

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt (hunk 3 above)
Comment (nikitamarysolomanpvt) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt:210]: +1

File: model/src/main/proto/exploration.proto (hunk 6 above)
Comment (nikitamarysolomanpvt) [reply model/src/main/proto/exploration.proto]: +1

File: model/src/main/proto/exploration.proto [hunk 8]
```diff
@@ -108,14 +108,29 @@ message Solution {
...
+  bool isNegative = 4;
+  string correct_answer = 5;
+}
 // Structure for a single hint
 // Maps from: data/src/main/java/org/oppia/data/backends/gae/model/GaeHint.kt
 message Hint {
   SubtitledHtml hint_content = 1;
+  bool hint_is_revealed = 2;
```
Comment (nikitamarysolomanpvt) [inline model/src/main/proto/exploration.proto:134]: Please comment the variable details.

File: model/src/main/proto/exploration.proto [hunk 9]
```diff
@@ -108,14 +108,29 @@ message Solution {
...
+  State state = 3;
```
Comment (nikitamarysolomanpvt) [inline model/src/main/proto/exploration.proto:136]: ditto

**Review by nikitamarysolomanpvt — APPROVED** [review]
Some nit suggestions PTAL

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt (hunk 1 above)
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Done.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt (hunk 2 above)
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt:216]: Done.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt (hunk 3 above)
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt:210]: Added statement inside it.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt (hunk 4 above)
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt:203]: Done.

File: model/src/main/proto/exploration.proto (hunk 6 above)
Comment (veena14cs) [reply model/src/main/proto/exploration.proto]: done.

File: model/src/main/proto/exploration.proto (hunk 8 above)
Comment (veena14cs) [reply model/src/main/proto/exploration.proto:134]: Done.

File: model/src/main/proto/exploration.proto (hunk 9 above)
Comment (veena14cs) [reply model/src/main/proto/exploration.proto:136]: Done.

File: model/src/main/proto/exploration.proto (hunk 5 above)
Comment (veena14cs) [reply model/src/main/proto/exploration.proto:112]: Done.

**Comment by veena14cs** [comment]
@rt4914 and @nikitamarysolomanpvt I have made changes related to revealing of hints and solution and also added testcases. PTAL.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt [hunk 10]
```diff
@@ -401,19 +478,44 @@ class ExplorationProgressController @Inject constructor(
...
+    internal fun computeHintForResult(currentState: State, hintIsRevealed: Boolean, hintIndex: Int): Hint {
+      val hintBuilder = Hint.newBuilder()
+        .setHintIsRevealed(hintIsRevealed)
+        .setHintContent(currentState.interaction.getHint(hintIndex).hintContent)
+        .setState(currentState)
+
+      return hintBuilder.build()
+    }
+
+
```
Comment (rt4914) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Remove this extra blank line

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt [hunk 11]
```diff
@@ -401,19 +478,44 @@ class ExplorationProgressController @Inject constructor(
       }
       return answerOutcomeBuilder.build()
     }
+
+    /** Returns an [Hint] based on the current state and revealed [Hint] from the learner's answer. */
+    internal fun computeHintForResult(currentState: State, hintIsRevealed: Boolean, hintIndex: Int): Hint {
+      val hintBuilder = Hint.newBuilder()
```
Comment (rt4914) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Update:
```
return Hint.newBuilder()
        .setHintIsRevealed(hintIsRevealed)
        .setHintContent(currentState.interaction.getHint(hintIndex).hintContent)
        .setState(currentState)
        .build()
```

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt [hunk 12]
```diff
@@ -401,19 +478,44 @@ class ExplorationProgressController @Inject constructor(
...
+        .setHintContent(currentState.interaction.getHint(hintIndex).hintContent)
+        .setState(currentState)
+
+      return hintBuilder.build()
+    }
+
+
+    /** Returns an [Solution] based on the current state and revealed [Solution] from the learner's answer. */
+    internal fun computeSolutionForResult(currentState: State, solutionIsRevealed: Boolean): Solution {
+      val solutionBuilder = Solution.newBuilder()
```
Comment (rt4914) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Update:
```
      return Solution.newBuilder()
        .setSolutionIsRevealed(solutionIsRevealed)
        .setAnswerIsExclusive(currentState.interaction.solution.answerIsExclusive)
        .setCorrectAnswer(currentState.interaction.solution.correctAnswer)
        .setExplanation(currentState.interaction.solution.explanation)
        .build
```

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt [hunk 13]
```diff
@@ -149,6 +151,79 @@ class ExplorationProgressController @Inject constructor(
...
+            hintIndex
+          )
+          explorationProgress.stateDeck.pushStateForHint(state, hintIndex)
+
+        } finally {
+          // Ensure that the user always returns to the VIEWING_STATE stage to avoid getting stuck in an 'always
+          // showing hint' situation. This can specifically happen if hint throws an exception.
+          explorationProgress.advancePlayStageTo(PlayStage.VIEWING_STATE)
+        }
+
```
Comment (rt4914) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Remove this blank line

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt [hunk 14]
```diff
@@ -149,6 +151,79 @@ class ExplorationProgressController @Inject constructor(
...
+        }
+        check(explorationProgress.playStage != PlayStage.LOADING_EXPLORATION) {
+          "Cannot submit an answer while the exploration is being loaded."
+        }
+        check(explorationProgress.playStage != PlayStage.SUBMITTING_ANSWER) {
+          "Cannot submit an answer while another answer is pending."
+        }
+        lateinit var hint: Hint
+        try {
+
```
Comment (rt4914) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Remove this blank line

File: domain/src/test/java/org/oppia/domain/exploration/ExplorationProgressControllerTest.kt [hunk 15]
```diff
@@ -100,6 +102,12 @@ class ExplorationProgressControllerTest {
   @Mock
   lateinit var mockAsyncAnswerOutcomeObserver: Observer<AsyncResult<AnswerOutcome>>
 
+  @Mock
+  lateinit var mockAsyncHintObserver: Observer<AsyncResult<Hint>>
+
+  @Mock
+  lateinit var mockAsyncSolutionObserver: Observer<AsyncResult<Solution>>
```
Comment (rt4914) [inline domain/src/test/java/org/oppia/domain/exploration/ExplorationProgressControllerTest.kt]: Optional Nit: These can be written in single lines.
`@Mock lateinit var mockAsyncSolutionObserver: Observer<AsyncResult<Solution>>`

Ditto everywhere else.

File: domain/src/test/java/org/oppia/domain/exploration/ExplorationProgressControllerTest.kt [hunk 16]
```diff
@@ -100,6 +102,12 @@ class ExplorationProgressControllerTest {
   @Mock
   lateinit var mockAsyncAnswerOutcomeObserver: Observer<AsyncResult<AnswerOutcome>>
 
+  @Mock
+  lateinit var mockAsyncHintObserver: Observer<AsyncResult<Hint>>
```
Comment (rt4914) [inline domain/src/test/java/org/oppia/domain/exploration/ExplorationProgressControllerTest.kt]: Press `Control + Alt + L` - reformatting shortcut to fix the formatting.

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt (hunk 10 above)
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Done.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt (hunk 11 above)
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Done.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt (hunk 12 above)
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Done.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt (hunk 13 above)
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Done.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt (hunk 14 above)
Comment (veena14cs) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationProgressController.kt]: Done.

File: domain/src/test/java/org/oppia/domain/exploration/ExplorationProgressControllerTest.kt (hunk 15 above)
Comment (veena14cs) [reply domain/src/test/java/org/oppia/domain/exploration/ExplorationProgressControllerTest.kt]: Updated everywhere.

File: domain/src/test/java/org/oppia/domain/exploration/ExplorationProgressControllerTest.kt (hunk 16 above)
Comment (veena14cs) [reply domain/src/test/java/org/oppia/domain/exploration/ExplorationProgressControllerTest.kt]: Done.

**Review by nikitamarysolomanpvt — APPROVED** [review]
LGTM

**Review by nikitamarysolomanpvt — APPROVED** [review]
LGTM

**Comment by veena14cs** [comment]
> LGTM

Thanks @nikitamarysolomanpvt

**Pull request merged**